src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/block.rs

```diff
@@ -0,0 +1,784 @@
+                VHOST_USER_BLOCK_DRIVER_NAME.to_string(),
+            ),
+            timer_fd: Some(TimerFd::new().map_err(VirtIoError::IOError)?),
+            intr_evts: Arc::new(intr_evts),
+            curr_queues: init_queues,
+            id: VHOST_USER_BLOCK_DRIVER_NAME.to_string(),
+        })
+    }
+
+    fn reconnect_to_server(&mut self) -> VirtIoResult<Master> {
+        if !Path::new(self.vhost_socket.as_str()).exists() {
+            return Err(VirtIoError::InternalError);
+        }
+        let master = Master::connect(&self.vhost_socket, 1).map_err(VirtIoError::VhostError)?;
+
+        Ok(master)
+    }
+
+    // vhost-user protocol features this device supports
+    fn get_dev_protocol_features(&self) -> VhostUserProtocolFeatures {
+        let mut features = VhostUserProtocolFeatures::MQ;
+
+        // TODO: need to support INFLIGHT_SHMFD later.
+        features |= VhostUserProtocolFeatures::CONFIG
+            | VhostUserProtocolFeatures::CONFIGURE_MEM_SLOTS
+            | VhostUserProtocolFeatures::REPLY_ACK;
+        // | VhostUserProtocolFeatures::INFLIGHT_SHMFD;
+
+        features
+    }
+
+    fn setup_slave<AS, Q, R>(&mut self, handler: &VhostUserBlockHandler<AS, Q, R>) -> ActivateResult
+    where
+        AS: DbsGuestAddressSpace,
+        Q: QueueT + Send + 'static,
+        R: GuestMemoryRegion + Sync + Send + 'static,
+    {
+        let mut config = EndpointParam {
+            virtio_config: &handler.config,
+            intr_evts: handler.config.get_queue_interrupt_eventfds(),
+            queue_sizes: &self.queue_sizes,
+            features: self.get_acked_features(),
+            protocol_flag: 0,
+            dev_protocol_features: self.get_dev_protocol_features(),
+            reconnect: false,
+            backend: None,
+            init_queues: self.curr_queues,
+            slave_req_fd: None,
+        };
+        config.set_protocol_mq();
+
+        loop {
+            match self.endpoint.negotiate(&config, None) {
+                Ok(_) => break,
+                Err(VirtIoError::VhostError(VhostError::VhostUserProtocol(e))) => {
+                    if e.should_reconnect() {
+                        // fall through to rebuild the connection.
+                        warn!(
+                            "{}: socket disconnected while initializing the connnection: {}",
+                            self.id, e
+                        );
+                    } else {
+                        error!("{}: failed to setup connection: {}", self.id, e);
+                        return Err(VhostError::VhostUserProtocol(e).into());
+                    }
+                }
+                Err(e) => {
+                    error!("{}: failed to setup connection: {}", self.id, e);
+                    return Err(ActivateError::InternalError);
+                }
+            }
+
+            // Sleep for 100ms to limit the reconnection rate.
+            let delay = std::time::Duration::from_millis(100);
```

Comment:
Just curious, why the magic number 100?

Reply:
I don't know, but other places also limit reconnection rates in this way.